core/src/nucleus/ribosome/api/link_entries.rs

```diff
@@ -0,0 +1,75 @@
+use action::{Action, ActionWrapper};
+use agent::state::ActionResponse;
+use hash_table::links_entry::*;
+use json::ToJson;
+use nucleus::ribosome::api::{
+    runtime_allocate_encode_str, runtime_args_to_utf8, HcApiReturnCode, Runtime,
+};
+use serde_json;
+use std::sync::mpsc::channel;
+use wasmi::{RuntimeArgs, RuntimeValue, Trap};
+
+/// ZomeApiFunction::LinkEntries function code
+/// args: [0] encoded MemoryAllocation as u32
+/// Expected complex argument: LinkEntriesArgs
+/// Returns an HcApiReturnCode as I32
+pub fn invoke_link_entries(
+    runtime: &mut Runtime,
+    args: &RuntimeArgs,
+) -> Result<Option<RuntimeValue>, Trap> {
+    // deserialize args
+    let args_str = runtime_args_to_utf8(&runtime, &args);
+    let input: Link = match serde_json::from_str(&args_str) {
+        Ok(entry_input) => entry_input,
+        // Exit on error
+        Err(_) => {
+            // Return Error code in i32 format
+            return Ok(Some(RuntimeValue::I32(HcApiReturnCode::ErrorJson as i32)));
+        }
+    };
+
+    // Create AddLink Action
+    let action_wrapper = ActionWrapper::new(Action::AddLink(input));
+    // Send Action and block for result
+    let (sender, receiver) = channel();
+    ::instance::dispatch_action_with_observer(
+        &runtime.action_channel,
+        &runtime.observer_channel,
+        action_wrapper.clone(),
+        move |state: &::state::State| {
+            let mut actions_copy = state.agent().actions();
+            match actions_copy.remove(&action_wrapper) {
```

Comment:
Why is this remove()? We are working on a copy anyway, so this is an awkward way to test if action_wrapper was in...

Reply:
I copied this form `get.rs`

Reply:
Yeah, I saw this before when I've replaced this kind of observer code in the futures action refactor.. This PR gave me a place to complain about it ;)